clippy_dev/src/main.rs

```diff
@@ -75,10 +121,117 @@ fn main() {
                 update_lints(&UpdateMode::Change);
             }
         },
+        ("new_lint", Some(matches)) => {
+            create_new_lint(
+                matches.value_of("pass"),
+                matches.value_of("name"),
+                matches.value_of("category"),
+            );
+        },
         _ => {},
     }
 }
 
+fn project_root() -> Result<PathBuf, io::Error> {
+    let current_dir = std::env::current_dir()?;
+    for path in current_dir.ancestors() {
+        let result = std::fs::read_to_string(path.join("Cargo.toml"));
+        if let Err(err) = &result {
+            if err.kind() == io::ErrorKind::NotFound {
+                continue;
+            }
+        }
+
+        let content = result?;
+        if content.contains("[package]\nname = \"clippy\"") {
+            return Ok(path.to_path_buf());
+        }
+    }
+    Err(io::Error::new(ErrorKind::Other, "Unable to find project root"))
+}
+
+fn open_files(lint_name: &str) -> Result<(File, File), io::Error> {
+    let project_root = project_root()?;
+
+    let test_file_path = project_root.join(format!("tests/ui/{}.rs", lint_name));
+    let test_file = OpenOptions::new().write(true).create_new(true).open(test_file_path)?;
+
+    let lint_file_path = project_root.join(format!("clippy_lints/src/{}.rs", lint_name));
+    let lint_file = OpenOptions::new().write(true).create_new(true).open(lint_file_path)?;
+
+    Ok((test_file, lint_file))
+}
+
+fn to_camel_case(name: &str) -> String {
+    name.split('_')
+        .map(|s| [&s[0..1].to_uppercase(), &s[1..]].concat())
+        .collect::<String>()
+}
+
+fn create_new_lint(pass: Option<&str>, lint_name: Option<&str>, category: Option<&str>) {
+    let pass = pass.expect("`pass` argument is validated by clap");
+    let lint_name = lint_name.expect("`name` argument is validated by clap");
+    let category = category.expect("`category` argument is validated by clap");
+
+    match open_files(lint_name) {
+        Ok((mut test_file, mut lint_file)) => {
+            let pass_type = match pass {
+                "early" => "EarlyLintPass",
+                "late" => "LateLintPass",
+                _ => {
+                    eprintln!("`pass_type` should only ever be `early` or `late`!");
+                    return;
+                },
+            };
+
+            let camel_case_name = to_camel_case(lint_name);
+
+            test_file
+                .write_all(
+                    format!(
+                        "#![warn(clippy::{})]
+
+fn main() {{
+    // test code goes here
+}}
+",
+                        lint_name
+                    )
+                    .as_bytes(),
+                )
+                .unwrap();
+
+            lint_file
+                .write_all(
+                    format!(
+                        "use rustc::lint::{{LintArray, LintPass, {type}}};
+use rustc::declare_lint_pass;
+use rustc_session::declare_tool_lint;
+
+declare_clippy_lint! {{
```

Comment:
Can you also add the basic documentation structure here?
```suggestion
declare_clippy_lint! {{
    /// **What it does:**
    ///
    /// **Why is this bad?**
    ///
    /// **Known problems:** None.
    ///
    /// **Example:**
    ///
    /// ```rust
    /// // example code
    /// ```
```

Reply:
done